File: oracle-mock/oracle.py
from web3 import Web3
import json
import time
import os
from dotenv import load_dotenv
from web3.middleware import geth_poa_middleware

# Load environment variables
load_dotenv()

PRIVATE_KEY = os.getenv('PRIVATE_KEY')
OWNER_ADDRESS = Web3.to_checksum_address(os.getenv('OWNER_ADDRESS'))
AUTOQUANT_ADDRESS = Web3.to_checksum_address(os.getenv('AUTOQUANT_ADDRESS'))
USDT_ADDRESS = Web3.to_checksum_address(os.getenv('ASSET'))

# Web3 provider (BNB Testnet)
w3 = Web3(Web3.HTTPProvider('https://data-seed-prebsc-1-s1.binance.org:8545/'))
w3.middleware_onion.inject(geth_poa_middleware, layer=0)

if not w3.is_connected():
    raise Exception("Failed to connect to Web3 provider")

# Load ABI files
with open('../artifacts/contracts/AutoQuant.sol/AutoQuant.json') as f:
    AUTOQUANT_ABI = json.load(f)['abi']

with open('../artifacts/contracts/mockUSDT.sol/MockUSDT.json') as f:
    ERC20_ABI = json.load(f)['abi']

# Instantiate contracts
usdt = w3.eth.contract(address=USDT_ADDRESS, abi=ERC20_ABI)
autoquant = w3.eth.contract(address=AUTOQUANT_ADDRESS, abi=AUTOQUANT_ABI)

def send_tx(tx):
    try:
        tx['nonce'] = w3.eth.get_transaction_count(OWNER_ADDRESS)
        tx['gasPrice'] = w3.eth.gas_price  # dynamic gas price

        signed_tx = w3.eth.account.sign_transaction(tx, PRIVATE_KEY)
        tx_hash = w3.eth.send_raw_transaction(signed_tx.raw_transaction)
        print(f"Sent tx: {tx_hash.hex()}")
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
        print("✅ Confirmed in block", receipt.blockNumber)
        return receipt
    except Exception as e:
        print(f"❌ Transaction failed: {str(e)}")
        return None

def mint_usdt(to_address, amount_wei):
    print("🟨 Minting USDT to wallet...")
    try:
        tx = usdt.functions.mint(to_address, amount_wei).build_transaction({
            'from': OWNER_ADDRESS,
            'gas': 200000,
        })
        return send_tx(tx)
    except Exception as e:
        print(f"❌ Mint failed: {str(e)}")

def approve_usdt(amount_wei):
    print("🟨 Approving USDT...")
    try:
        tx = usdt.functions.approve(AUTOQUANT_ADDRESS, amount_wei).build_transaction({
            'from': OWNER_ADDRESS,
            'gas': 100000,
        })
        return send_tx(tx)
    except Exception as e:
        print(f"❌ Approve failed: {str(e)}")

def deposit_usdt(amount_wei):
    print("🟨 Depositing to AutoQuant...")
    try:
        tx = autoquant.functions.deposit(amount_wei).build_transaction({
            'from': OWNER_ADDRESS,
            'gas': 200000,
        })
        return send_tx(tx)
    except Exception as e:
        print(f"❌ Deposit failed: {str(e)}")

def execute_buy():
    print("🟨 Executing BUY signal...")
    try:
        tx = autoquant.functions.executeSignal("BUY").build_transaction({
            'from': OWNER_ADDRESS,
            'gas': 400000,
        })
        return send_tx(tx)
    except Exception as e:
        print(f"❌ BUY failed: {str(e)}")

def get_contract_balances():
    try:
        contract_stable_balance = usdt.functions.balanceOf(AUTOQUANT_ADDRESS).call()
        user_balance = autoquant.functions.userStableBalances(OWNER_ADDRESS).call()
        allowance = usdt.functions.allowance(OWNER_ADDRESS, AUTOQUANT_ADDRESS).call()

        print(f"🧾 Contract stablecoin balance: {w3.from_wei(contract_stable_balance, 'ether')} USDT")
        print(f"🧾 User balance in AutoQuant: {w3.from_wei(user_balance, 'ether')} USDT")
        print(f"🧾 Allowance: {w3.from_wei(allowance, 'ether')} USDT")
    except Exception as e:
        print(f"❌ Failed to fetch balances: {str(e)}")

if __name__ == "__main__":
    # Assuming your mockUSDT uses 18 decimals as per your contract, so 1 USDT = 1e18 wei
    amount = w3.to_wei(1, 'ether')  # 1 USDT

    mint_usdt(OWNER_ADDRESS, amount)
    time.sleep(5)  # wait for mint tx confirmation

    approve_usdt(amount)
    time.sleep(5)  # wait for approval

    get_contract_balances()
    time.sleep(2)

    deposit_usdt(amount)
    time.sleep(5)

    get_contract_balances()
    time.sleep(2)

    execute_buy()

File: oracle-mock/oracle.js
const { ethers } = require("ethers");
require("dotenv").config();
const fs = require("fs");
const path = require("path");

// Load ABIs
const AUTOQUANT_ABI = JSON.parse(fs.readFileSync(path.join(__dirname, "../artifacts/contracts/AutoQuant.sol/AutoQuant.json"))).abi;
const ERC20_ABI = JSON.parse(fs.readFileSync(path.join(__dirname, "../artifacts/contracts/mockUSDT.sol/MockUSDT.json"))).abi;

// Env vars
if (!process.env.PRIVATE_KEY) {
    throw new Error("PRIVATE_KEY is not set in your .env file!");
}
const PRIVATE_KEY = process.env.PRIVATE_KEY.startsWith("0x")
    ? process.env.PRIVATE_KEY
    : "0x" + process.env.PRIVATE_KEY;
const OWNER_ADDRESS = process.env.OWNER_ADDRESS;
const AUTOQUANT_ADDRESS = process.env.AUTOQUANT_ADDRESS;
const USDT_ADDRESS = process.env.ASSET;

// Provider (BNB Testnet)
const provider = new ethers.JsonRpcProvider("https://data-seed-prebsc-1-s1.binance.org:8545/");
const wallet = new ethers.Wallet(PRIVATE_KEY, provider);

const usdt = new ethers.Contract(USDT_ADDRESS, ERC20_ABI, wallet);
const autoquant = new ethers.Contract(AUTOQUANT_ADDRESS, AUTOQUANT_ABI, wallet);

async function mintUSDT(to, amount) {
    console.log("🟨 Minting USDT to wallet...");
    try {
        const tx = await usdt.mint(to, amount);
        await tx.wait();
        console.log("✅ Minted USDT");
    } catch (e) {
        console.log("❌ Mint failed:", e.reason || e.message);
    }
}

async function approveUSDT(amount) {
    console.log("🟨 Approving USDT...");
    try {
        const tx = await usdt.approve(AUTOQUANT_ADDRESS, amount);
        await tx.wait();
        console.log("✅ Approved USDT");
    } catch (e) {
        console.log("❌ Approve failed:", e.reason || e.message);
    }
}

async function depositUSDT(amount) {
    console.log("🟨 Depositing to AutoQuant...");
    try {
        const tx = await autoquant.deposit(amount);
        await tx.wait();
        console.log("✅ Deposited to AutoQuant");
    } catch (e) {
        console.log("❌ Deposit failed:", e.reason || e.message);
    }
}

async function executeBuy() {
    console.log("🟨 Executing BUY signal...");
    try {
        const tx = await autoquant.executeSignal("BUY");
        await tx.wait();
        console.log("✅ BUY executed");
    } catch (e) {
        console.log("❌ BUY failed:", e.reason || e.message);
    }
}

async function getBalances() {
    try {
        const contractStableBalance = await usdt.balanceOf(AUTOQUANT_ADDRESS);
        const userBalance = await autoquant.userStableBalances(OWNER_ADDRESS);
        const allowance = await usdt.allowance(OWNER_ADDRESS, AUTOQUANT_ADDRESS);

        console.log(`🧾 Contract stablecoin balance: ${ethers.formatUnits(contractStableBalance, 18)} USDT`);
        console.log(`🧾 User balance in AutoQuant: ${ethers.formatUnits(userBalance, 18)} USDT`);
        console.log(`🧾 Allowance: ${ethers.formatUnits(allowance, 18)} USDT`);
    } catch (e) {
        console.log("❌ Failed to fetch balances:", e.reason || e.message);
    }
}

(async () => {
    const amount = ethers.parseUnits("1", 18); // 1 USDT

    await mintUSDT(OWNER_ADDRESS, amount);
    await new Promise(r => setTimeout(r, 5000));

    await approveUSDT(amount);
    await new Promise(r => setTimeout(r, 5000));

    await getBalances();
    await new Promise(r => setTimeout(r, 2000));

    await depositUSDT(amount);
    await new Promise(r => setTimeout(r, 5000));

    await getBalances();
    await new Promise(r => setTimeout(r, 2000));

    await executeBuy();
})();

console.log("PRIVATE_KEY loaded:", process.env.PRIVATE_KEY);